tensorflow/core/kernels/scatter_functor.h

```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
     // indices and params sizes were validated in DoCompute().
     const Index N = static_cast<Index>(indices.size());
     const Index limit = static_cast<Index>(params.dimension(0));
-    for (Index i = 0; i < N; i++) {
-      // Grab the index and check its validity.  Do this carefully,
-      // to avoid checking the value and grabbing it again from
-      // memory a second time (a security risk since it may change in between).
-      const Index index = ::tensorflow::internal::SubtleMustCopy(indices(i));
-      if (!FastBoundsCheck(index, limit)) return i;
-      // Copy last Ndim-1 dimensions of updates[i] to params[index]
-      scatter_op::internal::Assign<op>::Run(params.template chip<0>(index),
-                                            updates.template chip<0>(i));
+    unsigned long int num_locks, entries_per_lock;
+    // Duplicate entries need to be handled correctly.
+    // Multiple updates to the same index has to be serialized.
+    // To reduce the number of locks and the memory usage,
+    // we divide the whole index space into kMaxLocks regions
+    // with each lock serializing access to a region.
+    if (limit <= kMaxLocks) {
+      num_locks = limit;
+      entries_per_lock = 1;
+
+    } else {
+      num_locks = kMaxLocks;
+      entries_per_lock = (limit % kMaxLocks == 0) ? limit / kMaxLocks
+                                                  : (limit / kMaxLocks + 1);
```

Comment:
This can be more concise. Please also make each comment line fill the whole line (80 characters) before beginning a new line.
```c++
const Index num_locks = std::min(limit, kMaxLocks);
const Index entries_per_lock = (limit + kMaxLocks - 1) / kMaxLocks;
```